plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ThreadDumpingWatchdog.java

```diff
@@ -171,12 +201,110 @@ private void dumpEclipseLocks(StringBuilder sb, String linePrefix) {
       sb.append("\n").append(linePrefix);
       sb.append("\n").append(linePrefix).append("Eclipse Locks:");
       sb.append("\n").append(linePrefix).append(debugOutput.replace("\n", "\n" + linePrefix));
+      sb.append("\n").append(linePrefix).append("[").append(dumpingTime).append("]");
     } catch (SecurityException | IllegalArgumentException | ReflectiveOperationException ex) {
       sb.append("\n").append(linePrefix).append("Eclipse Lock information not available");
     }
   }
 
   /**
+   * Dump details on current jobs.
+   */
+  private void dumpEclipseJobs(StringBuilder sb, String linePrefix) {
+    Job[] jobs = Job.getJobManager().find(null);
+    if (jobs.length == 0) {
+      return;
+    }
+
+    Arrays.sort(jobs, new Ordering<Job>() {
+      @Override
+      public int compare(Job j1, Job j2) {
+        Preconditions.checkNotNull(j1);
+        Preconditions.checkNotNull(j2);
+        //@formatter:off
+        return ComparisonChain.start()
+            .compareTrueFirst(j1.isBlocking(), j2.isBlocking())
+            .compare(j2.getState(), j1.getState()) // descending order so RUNNING is first
+            .compare(j1.getPriority(), j2.getPriority())
+            .compareTrueFirst(j1.isUser(), j2.isUser())
+            .compareTrueFirst(j1.isSystem(), j2.isSystem())
+            .compare(j1.getRule(), j2.getRule(), Ordering.usingToString().nullsLast())
+            .result();
+        //@formatter:on
+      }
+    });
+
+    sb.append("\n").append(linePrefix).append(jobs.length + " jobs:");
+    for (int index = 0; index < jobs.length; index++) {
+      Job job = jobs[index];
+      dumpJob(sb, linePrefix, job, job.getThread());
+    }
+
+    // Try to dump ThreadJobs, which are threads that access ISchedulingRules
+    try {
+      Object implicitJobs =
+          ReflectionUtil.getField(Job.getJobManager(), "implicitJobs", Object.class);
+      Map<?, ?> threadJobs =
+          ReflectionUtil.getField(implicitJobs, "threadJobs", Map.class); /* <Thread,ThreadJob> */
+      if (!threadJobs.isEmpty()) {
+        sb.append("\n").append(linePrefix).append(threadJobs.size() + " ThreadJobs:");
+        for (Entry entry : threadJobs.entrySet()) {
+          dumpJob(sb, linePrefix, (Job) entry.getValue(), (Thread) entry.getKey());
+        }
+      }
+    } catch (Exception ex) {
+      System.err.println("Unable to obtain JobManager.implicitJobs: " + ex);
+    }
+  }
+
+
+  private void dumpJob(StringBuilder sb, String linePrefix, Job job, Thread thread) {
+    String status;
+    switch (job.getState()) {
+      case Job.RUNNING:
+        status = "RUNNING";
+        break;
+      case Job.WAITING:
+        status = "WAITING";
+        break;
+      case Job.SLEEPING:
+        status = "SLEEPING";
+        break;
+      case Job.NONE:
+        status = "NONE";
+        break;
+      default:
+        status = "UNKNOWN(" + job.getState() + ")";
+        break;
+    }
+    Object blockingJob = null;
+    try {
```

Comment:
The actual use of `blockingJob`, which is an InternalJob, is used further below, out of scope.